aten/src/ATen/templates/RegisterFunctionalization.cpp

```diff
@@ -37,6 +38,25 @@ constexpr auto exclude_keys_for_meta_dispatch =
         c10::DispatchKey::Python
     });
 
+// Helper around at::has_internal_overlap.
+// The ATen util is used in hot-path eager mode: it's always fast,
+// but might return TOO_HARD sometimes.
+// During functionalization, we're ok taking a bit longer
+// to detect memory overlap.
+inline bool has_internal_overlap_helper(const at::Tensor t) {
+  auto has_overlap = at::has_internal_overlap(t);
+  if (has_overlap == at::MemOverlap::Yes) return true;
+  if (has_overlap == at::MemOverlap::No) return false;
+  auto strides = t.sym_strides();
+  auto sizes = t.sym_sizes();
+  for (size_t i = strides.size() - 1; i > 1; ++i) {
+    // Another relatively simple heuristic.
+    if (strides[i] == 1 && strides[i-1] < sizes[i]) return true;
```

Comment:
cc @ezyang I'm a bit worried that this will add more horrible guards for inductor to sift through, wdyt?